Set temporary env whenever Julia is started with `--project=@temp`

ven-k opened the pull request:
This makes it convenient to start and run Julia in a temporary environment. This brings `Pkg.activate(; temp=true)` functionality to cmdline.

- Update relevant docs.
- Add tests for `--project=@temp` and `JULIA_PROJECT="@temp"`.

vtjnash commented:
Similar to https://github.com/JuliaLang/julia/pull/49061 (@KristofferC), and discussion there of whether we should spell this `--temp` to exactly match the Pkg REPL behavior

fingolfin commented:
This could breaks existing scripts, though. How about using `@temp` (similar to how there is `@.`)

ericphanson commented:
https://github.com/JuliaLang/julia/pull/50864 uses `@` as well

ven-k committed:
feat: set temporary env whenever Julia is started with `--project=@temp`

- Update relevant docs.
- Add tests for `--project=@temp` and `JULIA_PROJECT="@temp"`.

This makes it convenient to start and run Julia in a temporary environment. This brings `Pkg.activate(; temp=true)` functionality to cmdline.

fingolfin commented:
You have now changed it to `--project=--temp`. That's pretty unusual and fits into no other pattern, does it? Could you at least explain your rationale over the suggestion to use `@temp`, which would by the way also naturally fit in with the code in `load_path_expand`

ven-k commented:
The `@abc` is already used to activate shared env abc in the envdir(). So, avoiding `@temp` might be better.

Although `--project=--temp` looks unusual, it follows the `] activate --temp` pattern.

fingolfin reviewed (approved):
I find that argument compelling. Should this have a NEWS.md entry perhaps?

vtjnash reviewed (approved):
Yes, if someone could type a blurb for the NEWS file, this looks good to me

adienes commented:
I really appreciate the functionality but sorry must it be spelled `--project=--temp` ? I can't think of a single other CLI app I've ever used with the pattern `--arg1=--arg2`. I understand that within `Pkg` it is `] activate --temp`, but in this context it just seems extremely odd.

 I much prefer the original `--project=@temp` or something like `--temp-project`

vtjnash commented:
>  can't think of a single other CLI app I've ever used

It can be a somewhat common pattern in compiler drivers. E.g. `gcc -Wl,-z,undef` or `clang -mllvm`

ven-k committed:
doc: update NEWS.md about `--project=@temp`.

ven-k commented:
As conversations, here and in #49061, have gravitated toward `--project=--temp`, I've retained that format.

`--temp-project` looks very neat; but I'm not sure if a new argument that is a subset of another would be a good idea.

adienes commented:
reading through both these issue threads, I would not call it "gravitating," more just stalled after `--temp` was proposed. Just counting heads, I definitely see more contributors  proposing or preferring other options to `--temp` (by my count of the thumb wars, it's 6-3). and the suggestion for `--project=_` has 6x👍 from non-commenters

that being said, OSS is not a vote, and I am happy to be overruled. just wanted to put my feedback in that I find this syntax quite clunky.

@vtjnash that's a good point about the compiler drivers, but I strongly suspect that most users see Julia more like an app than they see it like a compiler

Keno commented:
`--temp` is technically a legal directory name also though. I think the whole flags vs file heuristics that clis do is a bit of an antipattern (e.g. why there's often things like `--` to separate them). I don't think it's any better than `--project-temp` though and I also think `--project=@temp` is fine (since `@` is already special; yes I know that it currently means the `temp` shared environment - I don't think it's a big issue, but if people are concerned, it could only have the new meaning if the temp shared environment doesn't exist).

mkitti commented:
How about an independent command line option? You would invoke this as `julia --temp`. It is an error to provide both `--project` and `--temp`. I could go for `--temp-project` as well. Having the value of an option use a `--` prefix looks strange to me and might even confuse some parsers.

IanButterworth commented:
--temp-project has the benefit that you could just type --temp because julia args only need to be complete enough to be unambiguous

Moelf commented:
in favor of `--temp`/`--temp-project` because `@temp` should be persistent just like any other `@abc` and `--arg1=--arg2` is weird among all the other command-line interface.

vtjnash commented:
> --temp is technically a legal directory name also though

Just as an observation for everyone, the pattern to force it to be a directory is to prepend `./` to make it `--project=./--temp` if you really must use a relative path to a folder of that name. Same with `@` being alternatively specified as `--project=@./temp` or `--project=./@temp` to force alternative meanings

jishnub commented:
Personally, I would prefer if this flag starts with `--project` so that it is easy to club it with the other forms in which `--project=` is used. From this perspective, I would prefer `--project=@temp` as it's the most visually similar form, but `--project-temp` also looks fine. `--temp-project` departs from this pattern, and even though it's easier to read than `--project-temp`, I would prefer consistency. The shortened form `julia --temp` looks cryptic to me, as it's unclear that the `temp` refers to the project.

GunnarFarneback commented:
Whatever is chosen here, please make it easy and convenient to use. Being able to write `julia --temp` is by far the best option I've seen suggested in this respect.

(As a side note, on international keyboards `@` tends to be tucked away, e.g. as `AltGr+2` on my keyboard.)

adienes commented:
> @temp should be persistent just like any other @abc

persistent-ly empty, so it kind of is

_something_ is going to have to be special-cased, and I think it's least surprising if that's behind an `@` symbol --- the universal Julian mark of "non-standard stuff be here"

and I know it has been mentioned a few times so this is a redundant observation, but nonetheless I think it's going unappreciated that `--project=@scriptdir` has _been already chosen_ to special case an `@keyword` project to be something other than a persistent shared environment

ericphanson commented:
> has been already chosen to special case an @keyword project to be something other than a persistent shared environment

It was chosen but not really by consensus; there was no discussion in that PR about the clash before it was merged, and I'm not sure that issue was even known (I brought it up post-merge). It seems like _a way_ to do it but it doesn't really seem clear that it's the best way, and it also doesn't seem totally to late to do something better, as it hasn't been released yet.

adienes commented:
that's fair, but I just mean the two args should not live by different rules (in fact, it seems cleaner to me if `scriptdir` and `temp` are accessible in similar ways, since they both special-case the `--project` parameter)

Moelf commented:
>persistent-ly empty, so it kind of is

That's not how persistent is usually defined in computing science realm I think. 

https://en.m.wikipedia.org/wiki/Persistence_(computer_science) summarize it as:

>refers to the characteristic of state of a system that outlives (persists more than) the process that created it.

So in this case the temp environment with non empty stuff is created by the Julia process, and thus is not persistent.

ericphanson commented:
From an interactive user point of view, `--project=temp` and `--project=scriptdir` actually seems slightly better than `--project=@temp` or `--project=@scriptdir`, because shared environments are accessible globally, so those are always meaningful names, whereas `--project=temp` is only a clash when there's actually ` temp` (or `scriptdir`) directory in the current working directory. (And having a shared environment named either does seem like a reasonable thing to have). 

But from a tooling point of view, `--project=temp` seems worse because there is probably a lot of tools/CI scripts/etc that expect `--project=$relpath` to work for some large set of `relpath`s (that might not include ones starting with `@`, but probably does include ones like `temp` or `scriptdir`).

To me the idea of new CLI arguments like `--project-temp` and `--project-scriptdir` makes sense in that there's no clash at all, but then one needs to thread through a new CLI argument for each...

What about something like `@temp` but using new syntax to not clash as much (besides with unusually named local dirs, which was already a problem with the introduction of `@` for shared envs and seems to have worked fine in practice)? e.g. `--project=+temp` or `--project=!temp` or something like that?

StefanKarpinski commented:
I think that expressing this as `--project=@temp` and allowing `@temp` in `LOAD_PATH` to create a temporary directory would be a good approach here. That is a feature that came up in https://github.com/JuliaLang/Pkg.jl/pull/2669#issuecomment-884502870; I posted a suggestion on Slack and then @IanButterworth mentioned it on GitHub (thank you, Ian): the idea is to have a `LOAD_PATH` of `@:@temp:@v#.#:@stdlib` for @oxinabox's preferred workflow where he wants to load a package but not into the currently active project, nor into a persistent named environment. If we let `@temp` mean that, then her preferred workflow would become much easier to implement. I agree with @keno that the concern that `@temp` might already be a named environment isn't a big one. It's weird to have a persistent named environment called `temp`.

StefanKarpinski commented:
Oh, I also don't like the name `@scriptdir` much and totally missed the discussion of it. I would prefer @tecosaur's proposal of `@script` that behaves like `@.` looking for a project directory in the ancestors of the path to the currently running script file.

mkitti commented:
If we use `@temp` I suggest that we should check if a shared environment named `temp` exists. We can then either
1. Issue a warning and then just use the existing shared directory (backwards compatibility).
2. Issue an error.

I'm learning towards option 1. It would be similar to a deprecation warning.

Additionally, we could offer a helpful error message about referring to the shared environment using its absolute path.

Another issue I want to point out is that setting the environment variable `JULIA_PROJECT=@shared_env` does currently have the intended effect. In this we think the user is referring to a local directory named `@shared_env`.

mkitti commented on test/loading.jl:
```diff
@@ -683,7 +683,7 @@ end
 # Base.active_project when version directory exist in depot, but contains no project file
 mktempdir() do dir
-    vdir = Base.DEFAULT_LOAD_PATH[2]
+    vdir = Base.DEFAULT_LOAD_PATH[3]
```
Why did this change?

mkitti commented:
> ```julia
> julia> Pkg.activate(; shared = "temp")
> ┌ Error: `temp` is an invalid name for a shared environment.
> │ To create a temporary environment use `Pkg.activate(; temp = true)`
> └ @ Main REPL[10]:2
> ```

I thought the `shared` keyword only accepted a `Bool`?

Also, we really should check for an existing shared environment called `temp`.

mkitti commented on base/initdefs.jl:
```diff
@@ -131,21 +132,23 @@ end
 # a trailing empty entry in JULIA_LOAD_PATH which will be replaced with defaults.
 
-const DEFAULT_LOAD_PATH = ["@", "@v#.#", "@stdlib"]
+const DEFAULT_LOAD_PATH = ["@", "@temp", "@v#.#", "@stdlib"]
```
Wait. I thought @StefanKarpinski was saying that @oxinabox could set this in her startup.jl, but this is changing the default for everyone.

We do _not_ want to change the default. That would be a breaking change.

KristofferC commented:
https://github.com/JuliaLang/julia/pull/49061 FTW!

StefanKarpinski commented:
Modulo @mkitti's comment about not changing the default `LOAD_PATH`, this seems great and triage approves. 💟

ven-k commented:
The changes to `LOAD_PATH` are dropped here[^1].

[^1]:https://github.com/JuliaLang/julia/compare/f8e71219ede13e427855dc251ca790ee13f49fd9..75a85a627a09569f1cec50a93e7191707321d08b

mkitti commented on base/initdefs.jl:
```diff
@@ -238,7 +238,8 @@ function init_active_project()
         project === nothing ? nothing :
         project == "" ? nothing :
-        startswith(project, "@") ? load_path_expand(project) : abspath(expanduser(project))
+        startswith(project, "@") ? load_path_expand(project) :
+        abspath(expanduser(project))
```
Is this just a formatting change? From a formatting perspective, I think indentation of line 242 would be useful.

From a soruce control perspective, I would prefer no change since this does not seem necessary for the feature being added
.
```suggestion
        startswith(project, "@") ? load_path_expand(project) :
            abspath(expanduser(project))
```

ven-k replied:
Indenting just the line 242 wouldn't do, considering previous 2 lines are also the else-options for ?:
But I agree, this change isn't necessary. I've reverted that [here](https://github.com/JuliaLang/julia/compare/75a85a627a09569f1cec50a93e7191707321d08b..b203b09a46c7afebf23e8fdfa5793b2e081b71af).

mkitti commented:
Thank you for the changes. I think this is mergeable.

I remain concerned about the unusual case where there does exist a shared environment called "temp" that people may have accessed via `@temp` in the past. I could have easily created such an environment when debugging or testing some set of packages.

We should consider checking if such an environment exists and issuing a warning in that case. That said, it could be another pull request.

JeffBezanson commented:
Let's rebase and merge this.
